### ISSUE TITLE
Editor crashes every time I try to play using an Internal brain type

### ISSUE THREAD

**Issue by beardordie** [Unity-Technologies/ml-agents#918]
I just installed the newest ml-agents beta following the newest installation guides today, 06/25/2018. I had v0.2 beta working fine on another machine, so I know roughly how it's supposed to behave. What I am experiencing now is that all example scenes work fine on Player type brain and Heuristic type brain, but any time I set it to Internal type brain and use the provided bytes file for each example, the Unity Editor crashes upon pressing Play. I am new to crashes, so I'm not sure how to troubleshoot. I've attached the Editor log but I don't know how to read it for relevant information. I'm using Unity 2018.1.2f1
[Editor.log](https://github.com/Unity-Technologies/ml-agents/files/2135199/Editor.log)

**Comment by xiaomaogy** [comment]
@beardordie Did you follow the new documentation guide? Do you have the new TensorflowSharp plugin? Have you installed the new python packages? 

Also can you list out your detailed steps that leads to a crash?

**Comment by Setmaster** [comment]
I just finished the installation of the assets needed for ml-agents to work and I'm having the same issue. [Editor.log](https://gist.github.com/Setmaster/bf4a17621868e9eb0bdb7d458458aa0b)

**Comment by xiaomaogy** [comment]
@Setmaster Thanks for the video and log, but this information is not enough for us to help you. Please tell us the detailed steps to reproduce your error, specify things like what os you are using, which installation guide did you follow etc.

**Comment by xiaomaogy** [comment]
Running learn.py ModuleNotFoundError:No module named 'docopt' - Solved it by wiriting (ml-agents) C:\Users\vi7or\Documents\Repositories\ml-agents\python>python ./learn.py --run-id=run01 --train

Did you use py instead of python before?

**Comment by xiaomaogy** [comment]
Also after you click the play button in the editor, does the play button just get stuck there like that forever?  Have you made sure you are using Tensorflow 1.7.1 in your python environment, and used the latest version of TensorFlowSharp plugin in the basic guide?

**Comment by Setmaster** [comment]
I don't remember trying py before and I'm not sure what you mean by stuck, a few seconds after pressing play the editor will crash. [Here is a list of installed packages](https://gist.github.com/Setmaster/bfbb84d9287ba14d262d3f4dc4a1025b) which includes the correct version of Tensorflow. I downloaded the package again from [here](https://s3.amazonaws.com/unity-ml-agents/0.4/TFSharpPlugin.unitypackage) ,and Unity says there is nothing new to import, also I downloaded the package and installed it yesterday so I believe I'm using the latest version.

**Comment by beardordie** [comment]
Yes I followed the newest guide. Yes I had the tensorsharp plug-in
imported. yes I had all proper versions of packages as listed in the
documentation requirements file, and I tried multiple versions of
tensorflow, all cpu versions, and I tried with anaconda as well as without.
I think my processor must not be compatible with tensorflow. Intel core 2
extreme.

But that being said, does Python and tensorflow actually need to be
installed on the machine just to run already-trained models from bytes
files?

**Comment by xiaomaogy** [comment]
@beardordie No it doesn't require the python and tensorflow to be installed. But we haven't tested on the cpu you have. 

@Setmaster This is something I've never seen, I've tested our repo with the Windows 10 with Unity 2018.1, and the Internal Brain works without any crash. 

Can you guys build the Unity executable with the Internal Brain checked, then run the built executable in the command line and see what happens? Without any error message I am not able to even guess what's going on wrong here.....

**Comment by Setmaster** [comment]
@xiaomaogy By executable do you mean player? About the cmd, what parameter would be used for this? Also, I built a player with the brain set to internal anyway, and it crashed when executed. [Here is a copy of the player if it's useful.](https://drive.google.com/file/d/1Aq_2rUi9_dvc9-VCWYzLKN9g2GgIF-WF/view?usp=sharing)

**Comment by xiaomaogy** [comment]
Hi @Setmaster, By executable I mean the stuff you've provided here. 

I've tried to run your built executable provided above on my windows machine (Win 10), and it works without any crash.  To this point I'm pretty sure it is a machine specific things.  @beardordie Does this built executable work on your computer?

@Setmaster Is there any thing special about your computer? Have you tried this on any other computer?

**Comment by beardordie** [comment]
That built executable crashes on my computer with Intel Core 2 Extreme. I'm not surprised that this older processor is not working, but I am surprised that a computer with the specs xiaomaogy listed would have any trouble with it.

**Comment by beardordie** [comment]
For reference, the GPU in my Intel Core 2 Extreme PC (which crashes upon using internal brain) has an AMD Radeon HD 5800. Again, a much older computer, but it should be able to run an internal brain regardless of whether it supports all the tensorflow stuff to train new brains.

**Comment by Setmaster** [comment]
@xiaomaogy  The other computer was a Thinkpad notebook, I don't know the exact specs, but I presume all of them are different from mine.  @beardordie What's your CPU?

**Comment by Livenvh** [comment]
Hello, same issue.

My specs : Windows10 / Unity 2018.1.0f2
Old proc too : i920 (hyper threading desactivated for OC purpose)

**Comment by xiaomaogy** [comment]
@mmattar The windows machine we have is working, but for these people it seems that certain cpu specs will make the internal brain crash.

**Comment by Livenvh** [comment]
additionnal information, I have this message when installing TFSharpPlugin :
Unloading broken assembly Assets/ML-Agents/Plugins/Android/TensorFlowSharp.Android.dll, this assembly can cause crashes in the runtime 

And I upgraded Unity to 2018.2.0f2 but the problem persiste

**Comment by m4Ssa** [comment]
Hello, I'm also experiencing the same issue.

Specs: Windows10 / Unity 2018.1.0f / TF 1.7.1 / I7 Q740

**Comment by xiaomaogy** [comment]
@m4Ssa @Livenvh @beardordie Could you please try the older version of the TensorFlowSharp plugin available here (https://s3.amazonaws.com/unity-ml-agents/0.3/ML-AgentsWithPlugin.unitypackage), with the TF 1.4.0 (change requirement.txt and then `pip install .` to change the tensorflow back to 1.4.0)? If the editor stops crashes with the older TensorFlowSharp plugin, then I will try to update this plugin and see if that can fix the problem. Right now I don't have a windows machine that will crash with the steps you guys described, so I am not able to find a solution for this.

**Comment by beardordie** [comment]
I'm not pursuing ML on this machine anymore, sorry. If I get a day to kill
some time with, I may give it a shot and update this thread.

**Comment by beardordie** [comment]
I would be happy to test run builds, though, if that's helpful, since
presumably the crash would happen on this machine even running someone
else's game/program if it uses internal brain. I just don't have Python
installed at all anymore and don't have time/plans to install it again here.

**Comment by xiaomaogy** [comment]
@beardordie Actually python is not related to this crash, so you don't need to install it to test it. If you have time to test, that would be really helpful. Thanks in advance.

**Comment by Livenvh** [comment]
@xiaomaogy 
I had test the old TensorFlowSharp (on 2018.2)
=> The type or namespace name `CommunicatorParameters' could not be found (in RpcCommunicator and SocketCommunicator scripts

**Comment by xiaomaogy** [comment]
@Livenvh How did you test it? It seems that some of your c# script has been changed. Are you sure all of your .cs scripts inside Assets/ML-Agents folder are in sync with the v0.4 master, and only the TensorFlowSharp plugin has been switched to the older version?

Also 2018.2 might not work, please use 2018.1.

**Comment by Livenvh** [comment]
juste finish to test 
new 2018.1 project 
+ fresh ml-agent 0.4 (juste unzip gitub version and adapt player setings)
+ TensorFlowSharp you link above
=> The type or namespace name `CommunicatorParameters' could not be found (in RpcCommunicator and SocketCommunicator scripts (same than 2018.2)

**Comment by xiaomaogy** [comment]
@Livenvh Did you run `pip install .` in the ml-agents/python folder? I will try to run this myself tomorrow I guess....

**Comment by m4Ssa** [comment]
@xiaomaogy 
When i try to use the older TFSharp version im getting an error in Unity:
`Assets/ML-Agents/Scripts/RpcCommunicator.cs(23,9): error CS0246: The type or namespace name `CommunicatorParameters' could not be found. Are you missing an assembly reference?
`
I've build my Tensorflow without Grpc support since it didnt work with Grpc enabled. Don't know if thats a related problem though.

**Comment by Livenvh** [comment]
@xiaomaogy Yes I ran pip install . but always same message

**Comment by m4Ssa** [comment]
@xiaomaogy 
Ok so I opend a fresh project in 2018.1, didnt download ml-agents 0.4 but used the 0.3 version + TFSharp plugin from the package you posted earlier and now the internal brain works without a crash for me. Still having the 'CommunicatorParameters' error on 0.4 though.

**Comment by jjjuande** [comment]
It seems that Unity 2018.2 doesn't trust TensorFlowSharp.Android.dll so it's unloaded when the Unity Platform target is set to Android. And with that .dll unloaded, the projects won't run in the Editor when the platform target is Android. They run fine on an Android device or in the Editor when the platform target is set to anything else than Android. (e.g. Windows)

When the project is loaded with TFSharp installed:
Unloading broken assembly Assets/ML-Agents/Plugins/Android/TensorFlowSharp.Android.dll, this assembly can cause crashes in the runtime

When the project runs with Platform target set to Android:
TypeLoadException: Could not find method due to a type load error
Brain.InitializeBrain (Academy aca, Communicator communicator) (at Assets/ML-Agents/Scripts/Brain.cs:209)

**Comment by bmobear** [comment]
Same issue here. Brain type Player/Heuristic/External work fine. Unity crashes when the play button is clicked and the brain type is set to **Internal**.

Specs: Ubuntu 16.04 64-bit, Intel Core i7-6850K, Python 3.5.2, tensorflow 1.7.1, Unity 2018.2.0b2, ml-agent 0.4

**Comment by Pyroevil** [comment]
Same issue for me too with all examples scenes provided with the toolkit ( v0.4b ).
Training (external) work just fine but if I want to check the result in internal mode , Unity just close after few seconds. It's crash with my bytes files but with the bytes files provided with the toolkit too. I delete and restart the project from start many times. Trying the master branch or the last release ( v0.4b ).

Because I thinking it's not directly related to my tensorflow installation and my trained bytes files , I give it a try with the TensorFlowSharp v0.3 instead of v0.4. Now it's working but not with all samples scenes. Only with one don't having "Discrete" visualisation or action vectors. " Continuous type work ". Discrete one give me error. This is only true with my bytes files. Bytes files provided with the examples work fine with TensorFlowSharp v0.3. So I get stuck to only look pre-trained files come with examples and cannot see the results of my own experiements.

My specs:
TensorFlow 1.7.1 (compiled myself without AVX)
Python 3.5.1
Windows 10 Family 64x
Laptop Asus X553MA
Intel Pentium CPU N3540
Unity 2018.2.0f2
ML Agents ToolKit v0.4b
TensorFlowSharp v0.4 and TensorFlowSharp v0.3
8Go of rams

**Comment by Liven28** [comment]
Any news about that bug?

**Comment by xiaomaogy** [comment]
@Liven28 We are still not sure what is causing this bug, it works on our windows test machine so we are still not able to reproduce it.

**Comment by xiaomaogy** [comment]
@Pyroevil The error message you posted is saying the bytes file you generated is using a different tensorflow version than you place you are using it. If you want to try ml-agents v0.3, then you might want to switch all of them (including the tensorflowSharp plugin to v0.3, the tensorflow version to 1.4).

**Comment by xiaomaogy** [comment]
@jjjuande The problem you mentioned is a different issue, the TensorFlow.Android.dll file is showing the error message, but it is not the cause for the crash.

**Comment by xiaomaogy** [comment]
@m4Ssa So if you want to try ml-agents v0.3, switch all of them to v0.3, the communicator parameter error might be due to the protobuf file not compatible.

**Comment by Liven28** [comment]
I had just reinstall windows10 Family 64x (with just windows and graphic drivers up to date, unity and ml-agent configured).
Same problem persist (crash on play internal brain).

Configuration : i7 920 / GTX 1060 / 12 Go Ram / Gigabyte GA-X58A-UD7
(no overclocking)

I tried on  2017.4 - 2018.1 - 2018.2 - 2018.3b
with Python 3.5.1, tensorflow 1.7.1,  toolkit v0.4 / v0.5

**Comment by Gaby10** [comment]
Same problem here, everything works... except when I chose internal. It closes the screen as soon as  I press play.

Using 2018.2.9f1 on Windows 10, CPU (not using GPU) Intel. Let me know if you want more info or testing,

**Comment by xiaomaogy** [comment]
This is not something we can solve right now due to the reason I mentioned earlier. In v0.6 (which will be released in a few weeks)we will change the way internal brain works (It will be a scriptable object instead of a gameobject, and it will be called Learning Brain). If you guys want to try you can check this PR https://github.com/Unity-Technologies/ml-agents/pull/1250.

**Comment by Liven28** [comment]
@xiaomaogy ok very interesting. I cross fingers.
If you need to test something, you know where I am.

**Comment by codemode89** [comment]
same here.
Windows 10 Pro 64 bit, AMD Phenom(tm) || x4 965.
Training works fine, also in the editor. But if i click on play, with an internal brain, unity instantly closes.

**Comment by destructor465** [comment]
Same, internal brains not working, on play unity closes instantly, while training works without a problem.

Specs:
Windows 10 Pro 64-bit (Build 17763)
Intel Pentium G4620
Unity 2018.2.13f1

**Comment by destructor465** [comment]
> My laptop has AMD A8-3520m processor, which without AVX support. Build of @Setmaster doesn`t works.

My CPU doesn't support AVX too, I will need to try you way.

**Comment by Liven28** [comment]
@xiaomaogy 
Hello, my i7 920 doesn't support AVX, I tried kudyk solution (ml_agent v0.5 / TensorFlowSharp from v0.3 ml_agent) et it seems to work!

I just tried 3Dball scene with internal brain and Unity didn't crash.
I didn't make more tests, waiting for official news about the viability of this solution.

**Comment by destructor465** [comment]
kudyk solution works, but some examples throws long exceptions, one of them:
`
TFException: NodeDef mentions attr 'output_dtype' not in Op<name=Multinomial; signature=logits:T, num_samples:int32 -> output:int64; attr=seed:int,default=0; attr=seed2:int,default=0; attr=T:type,allowed=[DT_FLOAT, DT_DOUBLE, DT_INT32, DT_INT64, DT_UINT8, DT_INT16, DT_INT8, DT_UINT16, DT_HALF]; is_stateful=true>; NodeDef: multinomial_3/Multinomial = Multinomial[T=DT_FLOAT, output_dtype=DT_INT64, seed=670408, seed2=108](dense_3/MatMul, multinomial_3/Multinomial/num_samples). (Check whether your GraphDef-interpreting binary is up to date with your GraphDef-generating binary.).
`
Working examples: 3D Ball, Bouncer, Crawler, Reacher, Tennis and Walker.

**Comment by flurrux** [comment]
kudyk solution works for me too

**Comment by Liven28** [comment]
is the v0.7 solve the problem?

**Comment by kudyk** [comment]
Hi everyone, v0.7 works with tensorflow 1.7.0 from [here](https://github.com/fo40225/tensorflow-windows-wheel/tree/master/1.7.0/py36/CPU/sse2).

**Comment by Liven28** [comment]
Does it meen that tensorflow 1.7.0 now worsk with non AVX CPUs (and we don't need to use 1.4.0 any more)
or that v0.7 doesn't work with 1.4.0 any more and non AVX CPUs can't use lm-agent ?

**Comment by kudyk** [comment]
@Liven28 Tensorflow that i use form the link above is third-party, unofficial, built with sse2 support by @fo40225 user. He have build of 1.7.1 too, but only for cuda gpu, and it doesn`t work for me.

**Comment by xiaomaogy** [comment]
Since we've switched from TensorFlowSharp to Barracuda, this issue is no longer relevant. I will close it for now. Feel free to open if you want to discuss more.
